### PR TITLE
image-customize: Add SRPM build/install with --build

### PR DESCRIPTION
**Pull request by martinpitt**
This is necessary at least for testing Cockpit on RHEL 8, where it currently
needs to do some special hacks to produce two source rpms with half of the
binary packages. This cannot be represented with a tarball build.

It might also be generally useful for some general development.

----

Tested locally with building cockpit's srpm into rhel-9-0 image.

### CODE REVIEW

**Comment by martinpitt** [comment]
Rebased. This was clean, but for good measure I re-ran all the tests in https://github.com/cockpit-project/cockpit-podman/pull/901 to be really sure.

**Comment by martinpitt** [comment]
The podman PR landed, so we can now regression-test that right here. Triggered.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
No hard no here, but as usual, some possible improvements to consider.

File: image-customize [hunk 1]
```diff
@@ -88,18 +88,23 @@ class InstallAction(ActionBase):
...
-    '''Build and install distribution package(s) from dist tarball'''
+    '''Build and install distribution package(s) from dist tarball or source RPM'''
 
     @staticmethod
     def execute(machine_instance, source):
         # upload the tarball or srpm
         sourcename = os.path.basename(source)
+
+        if source.endswith(".src.rpm"):
+            version = None  # srpm knows its own version
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:99]: Yes, that's been on my todo list since the original PR that introduced `--build`. c-podman/machine/starter-kit don't actually produce a .tarball file, so we can't use that -- but it seems easier to just update debian/changelog and PKGBUILD right away there, then I can simplify the `else:` branch here and remove the seddery from `build_{arch,deb}`.

[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:99]: I sent https://github.com/cockpit-project/cockpit-podman/pull/909 for this. Once that gets approved and lands, I'll apply the same thing to cockpit-machines, then I'll drop the VERSION/SOURCE seddery here.

[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:99]: And https://github.com/cockpit-project/cockpit-machines/pull/557 for exactly the same change for machines.

File: image-customize [hunk 2]
```diff
@@ -139,22 +144,26 @@ class BuildAction(ActionBase):
...
-    def build_rpm(machine, vm_source, version):
+    def build_rpm(machine, vm_source, _version):
         mock_opts = ''
         if opt_verbose:
             mock_opts += ' --verbose'
         if opt_quick:
             mock_opts += ' --nocheck'
 
         # build source package
-        machine.execute(f'rpmbuild --define "_srcrpmdir /var/tmp/build" -ts "{vm_source}"')
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:150]: I did this in [this commit](https://github.com/martinpitt/bots/commit/a33e2adfb680f5974e655a889ab7a6776228901a). It's nicer for RPM, not much difference for Debian, but looks awful for arch. One can't run the build as root (`ERROR: Running makepkg as root is not allowed.`), and having this extra dedent/su wrapper/extra escape of `\` makes that shell snippet hard to read and copy&paste. I will not PR this version, just wanted to link it here for reference.

I'd really like to keep the chown for arch. I see nothing dangerous with it.

File: image-customize (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:99]: Once this lands, we can [land this commit](https://github.com/martinpitt/bots/commit/871ac599263ae6f301f42c667a556547112c60a1) to drop the seddery from image-customize.

File: image-customize (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:150]: Done in #2902. I did a compromise for arch which keeps the script legibility but avoids the textwrap and quoting.

File: image-customize (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply image-customize:99]: Done in #2902, I'll rebase this on top of that.

File: image-customize (hunk 2 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply image-customize:150]: I'm not sure why we need to do the source-fiddling in the arch case.  I actually have nothing against chown: my original observation was merely that it doesn't make sense to `mkdir` in one place and `chown` in another: for the arch case, why don't we just do exactly as we were doing before, but with the addition of `mkdir` at the top?

File: image-customize [hunk 3]
```diff
@@ -140,14 +141,18 @@ class BuildAction(ActionBase):
         if opt_quick:
             mock_opts += ' --nocheck'
 
-        # build source package
-        machine.execute(f'''su builder -c 'rpmbuild --define "_topdir /var/tmp/build" -ts "{vm_source}"' ''')
```
Comment (allisonkarlitskaya) [inline image-customize]: Why not this:

```
   flag = 'r' if vm_source.endswith('.src.rpm') else 't'
   machine.execute(f'''... -{flag}s {vm_source} ...''')
```
?

Comment (martinpitt) [reply image-customize]: FTR: It may make the code smaller, but I find it harder to understand then -- we already have an srpm, and I'd like it to build the srpm exactly as given; I am not sure if rebuilding an srpm from an srpm gives you the exact same result, and also it just feels like a waste of time, and loss of predictability.

**Review by allisonkarlitskaya — APPROVED** [review]
We discussed this on IRC.  Summary:
 - the `rpmbuild -rs` thing is maybe a nice simplification to the script, but it introduces unnecessary moving parts in rebuilding the already-built SRPM, so let's keep the way we have here
 - in the end, the tricks with `mkbuild.sh` in the arch case are pretty clean, so let's keep them.

tl;dr: approved!  Thanks for all the back and forth.  This looks really good now :)
